esp/logging/loggingagent/cassandraloggingagent/cassandralogagent.cpp

```diff
@@ -0,0 +1,621 @@
+
+        StringBuffer logID;
+        getLoggingTransactionID(logID);
+
+        CIArrayOf<CLogTable>& logTables = logGroup->getLogTables();
+        setKeySpace(logDB.str());
+        ForEachItemIn(i, logTables)
+        {
+            CLogTable& table = logTables.item(i);
+
+            StringBuffer cqlStatement;
+            if(!buildUpdateLogStatement(logRequestTree, logDB.str(), table, logID, cqlStatement))
+                throw MakeStringException(EspLoggingErrors::UpdateLogFailed, "Failed in creating SQL statement.");
+
+            if (getEspLogLevel() >= LogMax)
+                DBGLOG("CQL: %s\n", cqlStatement.str());
+
+            cassSession->connect();
+            executeSimpleStatement(cqlStatement);
+            cassSession->disconnect();
+        }
+        resp.setStatusCode(0);
+        return true;
+    }
+    catch (IException* e)
+    {
+        StringBuffer errorStr, errorMessage;
+        errorMessage.append("Failed to update log: error code ").append(e->errorCode()).append(", error message ").append(e->errorMessage(errorStr));
+        ERRLOG("%s", errorMessage.str());
+        e->Release();
+        resp.setStatusCode(-1);
+        resp.setStatusMessage(errorMessage.str());
+    }
+    return false;
+}
+
+CLogGroup* CCassandraLogAgent::checkLogSource(IPropertyTree* logRequest, StringBuffer& source, StringBuffer& logDB)
+{
+    if (logSourceCount == 0)
+    {//if no log source is configured, use default Log Group and DB
+        logDB.set(defaultDB.str());
+        return logGroups.getValue(defaultLogGroup.get());
+    }
+    source = logRequest->queryProp("Source");
+    if (source.isEmpty())
+        throw MakeStringException(EspLoggingErrors::UpdateLogFailed, "Failed to read log Source from request.");
+    CLogSource* logSource = logSources.getValue(source.str());
+    if (!logSource)
+        throw MakeStringException(EspLoggingErrors::UpdateLogFailed, "Log Source %s undefined.", source.str());
+
+    logDB.set(logSource->getDBName());
+    return logGroups.getValue(logSource->getGroupName());
+}
+
+void CCassandraLogAgent::getLoggingTransactionID(StringBuffer& id)
+{
+    CriticalBlock b(uniqueIDCrit);
+    id.set(transactionSeed.str()).append("-").append(++loggingTransactionCount);
+}
+
+bool CCassandraLogAgent::buildUpdateLogStatement(IPropertyTree* logRequest, const char* logDB, CLogTable& table, StringBuffer& logID, StringBuffer& cqlStatement)
+{
+    StringBuffer fields, values;
+    BoolHash handledFields;
+    CIArrayOf<CLogField>& logFields = table.getLogFields();
+    ForEachItemIn(i, logFields) //Go through data items to be logged
+    {
+        CLogField& logField = logFields.item(i);
+
+        StringBuffer colName = logField.getMapTo();
+        bool* found = handledFields.getValue(colName.str());
+        if (found && *found)
+            continue;
+
+        StringBuffer path = logField.getName();
+        if (path.charAt(path.length() - 1) == ']')
```

Comment:
libxml or xerces might provide xpath support so you don't have to do some of this yourself...